abci/app.go

```diff
@@ -0,0 +1,378 @@
+	resp.Value = resultBytes
+
+	return &resp, nil
+}
+
+func (app ForumApp) CheckTx(ctx context.Context, checktx *abci.RequestCheckTx) (*abci.ResponseCheckTx, error) {
+	// Parse the tx message
+	msg, err := model.ParseMessage(checktx.Tx)
+	if err != nil {
+		fmt.Printf("failed to parse transaction message checktx: %v\n", err)
+		return &abci.ResponseCheckTx{Code: CodeTypeInvalidTxFormat, Log: "Invalid transaction format"}, err
+	}
+	fmt.Println("Searching for sender ... ", msg.Sender)
+	u, err := app.state.DB.FindUserByName(msg.Sender)
+
+	if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+		fmt.Println("problem in check tx: ", string(checktx.Tx))
+		return &types.ResponseCheckTx{Code: CodeTypeEncodingError}, nil
+	}
+
+	if err == badger.ErrKeyNotFound {
+		fmt.Println("Not found user :", msg.Sender)
+	}
+	if u != nil && u.Banned {
+		return &types.ResponseCheckTx{Code: CodeTypeBanned, Log: "User is banned"}, nil
+	}
+	fmt.Println("Check tx success for ", msg.Message, " and ", msg.Sender)
+	return &types.ResponseCheckTx{Code: CodeTypeOK}, nil
+}
+
+// Consensus Connection
+// Initialize blockchain w validators/other info from CometBFT
+func (app ForumApp) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	for _, v := range req.Validators {
+		app.updateValidator(v)
+	}
+	appHash := make([]byte, 8)
+	binary.PutVarint(appHash, app.state.Size)
+	req.ConsensusParams.Abci.VoteExtensionsEnableHeight = 1
+	return &abci.ResponseInitChain{ConsensusParams: req.ConsensusParams, AppHash: appHash}, nil
+}
+
+func (app *ForumApp) PrepareProposal(_ context.Context, proposal *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	fmt.Println("entered prepareProp")
+
+	voteExtensionCurseWords := app.getWordsFromVe(proposal.LocalLastCommit.Votes)
+
+	// prepare proposal puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range proposal.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err == nil {
+			if err != nil {
+				proposedTxs = append(proposedTxs, tx)
+			}
+			// Adding the curse words from vote extensions too
+			if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+				proposedTxs = append(proposedTxs, tx)
+			} else {
+				banTx := model.BanTx{UserName: msg.Sender}
+				bannedUsersString[msg.Message] = struct{}{}
+				resultBytes, err := json.Marshal(banTx)
+				if err == nil {
+					finalProposal = append(finalProposal, resultBytes)
+				}
+			}
+		}
+	}
+	// Need to loop again through the proposed Txs to make sure there is none left by a user that was banned after the tx was accepted
+	for _, tx := range proposedTxs {
+		// there should be no error here as these are just transactions we have checked and added
+		msg, _ := model.ParseMessage(tx)
+		if _, ok := bannedUsersString[msg.Sender]; !ok {
```

Comment:
Just double checking... you're looking up `msg.Sender` here, but you set `msg.Message` above (line 176). I'm not sure I get it

Reply:
This is the extra check I explained in the call. This loop is done after all transactions have been processed. And what can happen is :
`Message: {sender:bob; message:hi}` -> OK, add to `proposedTx`
`Message: {sender:bob; message:curse_workd}` -> notOK , create `BanTx{username:Bob}` ; add this to `finalProposal` and add `bob` to `bannedUserString`.


Now, if we just appended `proposedTx` to `finalProposal` bobs first message would get in. But we want to ban bob and to not want his messages. So we make sure that this block will not contain any message from bob. And `bannedUserString` is used to verify that the sender of a message that is already in `proposedTx` does not exist in this map. If it does not exist, means the user was not banned here.

Reply:
Talked online, `msg.Sender` `msg.Message` are inconsistent